darwin/dataset/upload_manager.py

```diff
@@ -1,528 +1,613 @@
-        raise ValueError(f"Class name is {name}. Invalid value.")
-
-    payload = {
-        "annotation_type_ids": annotation_type_ids,
-        "cropped_image": cropped_image,
-        "dataset_id": dataset_id,
-        "description": description,
-        "metadata": metadata,
-        "name": name,
-    }
-
-    endpoint = "annotation_classes"
-    response = client.post(endpoint=endpoint, payload=payload, team=team, retry=True)
-    if "errors" in response:
-        raise ValueError(f"Response error: {response['errors']}")
-    return response
+class LocalFile:
+    def __init__(self, local_path, **kwargs):
+        self.local_path = Path(local_path)
+        self.data = kwargs
+        self._type_check(kwargs)
+
+    def _type_check(self, args):
+        self.data["filename"] = args.get("filename") or self.local_path.name
+        self.data["type"] =  args.get("type") or self._infer_type()  
+
+    def _infer_type(self):
+        extension = Path(self.data["filename"]).suffix
+        if is_image_extension_allowed(extension):
+            return "image"
+        if is_video_extension_allowed(extension):
+            return "video"
+        raise UnsupportedFileType(self.data["filename"])
+
+class UploadHandler:
+    def __init__(self, client, local_files, dataset_identifier):
+        self.client = client 
+        self.local_files = local_files
+        self.dataset_identifier = dataset_identifier
+
+        self.blocked_items, self.pending_items = request_upload(client, local_files, dataset_identifier)
+        print(self.blocked_items, self.pending_items)
+
+    @property
+    def pending_count(self):
+        return len(self.pending_items)
+
+    @property
+    def blocked_count(self):
+        return len(self.blocked_items)
+
+    def upload(self):
+        return _upload_files(self.client, self.local_files, self.pending_items, self.dataset_identifier.team_slug)
+    
+
+def request_upload(client: "Client", files: List[LocalFile], dataset_identifier):
+    upload_payload = { "items": [file.data for file in files] }
+    data = client.put(endpoint=f"/teams/{dataset_identifier.team_slug}/datasets/{dataset_identifier.dataset_slug}/data", payload=upload_payload, team=dataset_identifier.team_slug)
+    return data["blocked_items"], data["items"]
+
+# def add_files_to_dataset(client: "Client", files: List[LocalFile], dataset_slug: str, team_slug: str):
+#     upload_payload = { "items": [file.data for file in files] }
+#     data = client.put(endpoint=f"/teams/{team_slug}/datasets/{dataset_slug}/data", payload=upload_payload, team=team_slug)
+    
+#     print(data)
+
+#     if 'items' in data and data['items']:
+#         _upload_files(client, files, data['items'], team_slug)
+
+def _upload_files(client: "Client", files: List[LocalFile], items_pending_upload, team_slug: str):
+    file_lookup = { file.data["filename"]: file for file in files }
+    for item in items_pending_upload:
+        file = file_lookup.get(item['filename'])
```

Comment:
This is broken if a user uploads two files with the same name but different dest folders, since we would always match to just one of them.

Solutions:
1) introduce the request_id we talked about (Cons: fair bit of work)
2) use the dest path + filename as the key (Cons: make sure to normalize that path, and are we even getting the path back in the response? if not we need to add it!)